### DISCUSSION TITLE
[PYBD_SF2] Implementing an option to change to external antenna on bluetooth module

### DISCUSSION THREAD

**Discussion by jotamudo** [micropython/micropython#12368] [STM32 / Pyboard] [answered]
Hi guys :) 👋🏻
So I'm developing a project with a small company and we noticed that on the micropython port for the pyboard SF2 we're unable to change the antenna configuration (internal/external) without activating the network module via:
```python
import network

wl = network.WLAN(1)
wl.config(antenna = 1)
wl.activate(1) # <- seems really wasteful to activate the wifi module just to use the external antenna on the BT module
```
even though the switch should be made by just changing an GPIO state (WL_GPIO_1 connected to the RF switch).
I'm not seeing a clean way to do it besides doing a completely new module just to add a single method that does it, as copying the kwargs parsing done on the config function of the network module:
```c
switch (mp_obj_str_get_qstr(e->key)) {
    case MP_QSTR_antenna: {
        uint8_t buf[4];
        nw_put_le32(buf, mp_obj_get_int(e->value));
        cyw43_ioctl(self->cyw, CYW43_IOCTL_SET_ANTDIV, 4, buf, self->itf);
        break;
    }
```
is basically impossible without instantiating a self object of type `network_cyw43_obj_t` (that would need to come from somewhere or be statically created), which would make the `bluetooth` thereafter 100% dependent on the cyw4343 driver and modules. I tried looking into the driver code to see if there was a way to add such an option specifically for the cyw4343 port. I'm willing to make a PR to add the option and some documentation about it, as I suppose someone may also want the same.
Do you guys have any ideas/suggestions for such?

**Comment by jimmo** [comment]
@jotamudo I will check with Damien, but I think that to drive this GPIO pin, the WiFi interface needs to be active anyway (because it's the WiFi interface on the cyw43 that controls the GPIO).

In other words, you could put this code elsewhere, but `cyw43_ioctl` still relies on the cyw43 wifi interface being active.

It's not particularly wasteful - activating the interface doesn't cost any RAM, and you can also set the `pm` config option to put it into the lower power mode (e.g. `wl.config(pm=wl.PM_POWERSAVE)`).

**Comment by dpgeorge** [comment, marked as answer]
The antenna handling is a little complicated.  Basically there are two options:
1. Let the CYW4343 handle the antenna switching itself via antenna diversity.  It does this by measuring the signal strength on both antenna paths and automatically selecting the best one.  This is what the `CYW43_IOCTL_SET_ANTDIV` ioctl is used for, to enable antenna diversity.
2. Handle antenna selection yourself.  To do this you can simply control the WL_GPIO_1 pin from the main STM MCU itself.  In this case you don't need to enable wifi at all.

I think option 2 is exactly what you want.  To select the external antenna do:
```
machine.Pin("I11", machine.Pin.IN, machine.Pin.PULL_UP)
```
It's input mode in case the CYW4343 itself tries to drive that line.  Pull up selects the external antenna path.  Pull down will select the on-board chip antenna.

You must run that code after activating the BLE interface.

**Reply by jimmo** [reply]
@jotamudo FYI.

**Reply by jotamudo** [reply]
It does solve it indeed, I was afraid of controlling the pin via GPIO directly and having it change back later to another state in another piece of the firmware.
Out of curiosity, does this antenna diversity option have something to do with the firmware blob that's sent to the CYW4343 on the initialization?

**Comment by jotamudo** [comment]
Thanks for the help @dpgeorge and @jimmo !
